[WIP] user-management react module

Hyperkid123 opened the pull request:
This should not be included to Hammer release.

This PR replaces the whole user section in access control with React components.

## Changes
- State management via redux
- Client routing using react-router (also connected to tree-view-controller and toolbars)
- Using miq API where possible
- Data lazy loading (no server roundtrip)

## TO-DO

- [x] use bulk assignment for tagging users https://github.com/ManageIQ/manageiq-api/issues/459
- [x] lots of specs
- [x] code cleanup
- [x] ~disable tree navigation while adding/editing/tagging user~
- [x] hide toolbar while adding/editing/tagging user
- [x] delete dead code from OpsController and OpsRbac
- [x] add prop-type checking
- [x] add actions to group and role links in user detail
- [x] add current tags list to user detail
- [x] filter tag categories with no entries
- [x] update react-ui-components version after https://github.com/ManageIQ/react-ui-components/pull/77 , https://github.com/ManageIQ/react-ui-components/pull/78, https://github.com/ManageIQ/react-ui-components/pull/88 and https://github.com/ManageIQ/react-ui-components/pull/87 are merged
- [x] add flash messages on events
- [x] add error handling on server response
- [x] add pagination to user table depends on https://github.com/ManageIQ/react-ui-components/pull/78
- [x] w8 for https://github.com/ManageIQ/manageiq-api/pull/464 to get merged
- [x] add is admin checking: https://github.com/ManageIQ/manageiq-ui-classic/pull/4756

EDIT: most of the PR are test related files i did not write 3k+ lines of code

himdel commented on app/assets/javascripts/controllers/tree_view_controller.js:
```diff
@@ -8,8 +8,16 @@
     listenToRx(function(payload) {
+      if (payload.type === 'init-react-routing' && !vm.reactRouting) {
+        vm.reactRouting = !!payload.reactRouting;
+        vm.$scope.$apply;
```
missing `()`?

Hyperkid123 replied:
nice catch!

Hyperkid123 commented:
@ZitaNemeckova @lpichler

lpichler commented on app/controllers/ops_controller.rb:
```diff
@@ -91,6 +91,43 @@ def x_button
+
+  def get_category_entries
+    x = Classification.find_by_id(params[:cat_id])
+    render :json => x.entries.map { |entry| { :value => entry[:id], :label => entry[:description], :name => entry.tag[:name] } }
```
`entries.includes(:tag) `to avoid N+1 queires and you can use `.` notation  { :value => entry.id,...` to access attributes.

thanks

lpichler commented on app/controllers/ops_controller.rb:
```diff
@@ -91,6 +91,43 @@ def x_button
+  def get_category_entries_multi
+    category_entries = {}
+    params[:ids].map { |category_id|
+      cat = Classification.find_by_id(category_id)
```
I believe that this could be done in query
`Classification.where(:ids =>  params[:ids]).includes(:tags)....`

I see similar pattern in such queries maybe you will find how to unify it to one method and put it on model.

Hyperkid123 commented:
@dclarizio @martinpovolny @lpichler  i think it is ready for some clicking in the UI. You just need to add this PR  ManageIQ/manageiq-api#464  to your API plugin.

dclarizio commented:
@Hyperkid123 so tried to set everything up to test this.  Checked out manageiq-api repo, pr 464 branch.  Checked out manageiq-ui-classic repo, pr 4562 branch.  Have Gemfile.dev.rb overrides in to use the local sources.  Ran bin/update, successful.  Started up MIQ server, login, and get this:

![test1](https://user-images.githubusercontent.com/3433754/46373772-c41acf80-c643-11e8-908f-f437bca9cccd.png)

Which is the main Dashboard screen, but with the Users list view in the main area.

Regardless, I go onward to the Configuration/Access Control area to try to test the new form.  Users list view is good, but when I click on a user, I get:

![test2](https://user-images.githubusercontent.com/3433754/46373891-21af1c00-c644-11e8-9adf-a11c9c9a9749.png)

Not sure if I have something not set up right, so ping me when you get a chance.

Hyperkid123 commented:
@dclarizio ok the dashboard is purely for testing. I will obviously remove it when the time comes.
The second error is related to the API for the custom buttons events. It seems that the endpoint does not exists. It is in this PR https://github.com/ManageIQ/manageiq-api/pull/464. Did you override the manageiq-api plugin? I will check if something else did not change and let you know but i am confident that it works.

dclarizio commented:
@Hyperkid123 one more note, as above, if the tree is positioned on something other than the Users node (say the Region or Groups node), but I click on a user in the tree, the same thing happens.  The user is selected in the tree, but the right side shows the list of users.

Hyperkid123 commented:
@dclarizio thanks for the feed back i will try to resolve it ASAP

Hyperkid123 commented:
> Also, after adding or saving, I am positioned back on the list instead of on the record I just added or saved. If I cancel an edit, I am properly positioned back on the user I was viewing.

@dclarizio ok i know where is the "problem". The add/save routing is based on the previous location. If you are adding/editing from the list view, you will be redirected to the list view. If you are editing from user detail, you will be redirected back to the detail. 

So the correct behavior is to always redirect to the new/edited user? Even if you are editing from the list view?

dclarizio commented:
> So the correct behavior is to always redirect to the new/edited user? Even if you are editing from the list view?

I'm pretty sure that if you are on the list view when you select a user to edit, then you should return to the list view.  But if you are editing from the user summary screen, you should return to that.  As for Adds, I think it is probably ok to return to the list view as well (I may have been wrong there).

Hyperkid123 commented:
@dclarizio ok the routing issues should be fixed now

dclarizio commented:
@Hyperkid123 doing more testing.  The routing issues seem to be fixed, so that's good.

Not sure if this is related to this PR or just an overall multi-select drop down bug, but the group selection drop down list is not sorted, nor are the selected choices.  So, if I remove the 2 groups shown below and re-add them, they come back in reverse order and the Save button is still enabled.  Also, really hard to find them since the drop down list isn't sorted.

![test4](https://user-images.githubusercontent.com/3433754/46763323-551a2800-cc8e-11e8-8431-2da4bd2dc400.png)

Will do more testing as time permits.  BTW, hopefully you're looking at all the CC issues, as we can't merge with all that red.  :cry:

Hyperkid123 commented:
> So, if I remove the 2 groups shown below and re-add them, they come back in reverse order and the Save button is still enabled. Also, really hard to find them since the drop down list isn't sorted.

I will add sorting and searching for the multi select



> Will do more testing as time permits. BTW, hopefully you're looking at all the CC issues, as we can't merge with all that red.

Well about 60 of those errors are for test files... It is complaining about similar blocks of code and file size...

dclarizio commented:
@Hyperkid123 
> Does the order of the groups in selection matter?

Well, when I started the edit, they were in 1 order and after I removed them and added them back, the same choices were chosen but in a different order and the Save was enabled.  So, I guess it doesn't matter, as long as the control realizes when they are the same.  Perhaps sorting them makes sense anyway, just so it's easier to see what you have selected.

Hyperkid123 commented:
> Well, when I started the edit, they were in 1 order and after I removed them and added them back, the same choices were chosen but in a different order and the Save was enabled. 

Ok i will check it out, that is probably a bug, it should be always sorted alphabetically. The form control will be confused because it compares initial values with current state and if the order is not the same, the it will enable the buttons.

Thanks!

Hyperkid123 commented:
@dclarizio @himdel ok after a lot debugging hacking and discussing, we (me and @himdel) have decided that introducing client-side routing at this stage will create more problems than we have right now. Until other major changes are done to the app, adding client side routing to existing parts of the ui is not very beneficial. There is a ton of interactions between trees, toolbars and other that make this unnecessarily complicated (the code shows it) and not very maintainable.

I will extract all the components i have made simply put them inside the layout without the routing.
